File: bindings/c++/examples/hash_example.cpp
/**
 * @file hash_example.cpp
 * @brief QUAC 100 C++ SDK - Hash Operations Example
 * @copyright Copyright © 2025 Dyber, Inc. All Rights Reserved.
 *
 * Demonstrates hardware-accelerated hash functions.
 */

#include <quac100/quac100.hpp>
#include <iostream>
#include <iomanip>
#include <chrono>

using namespace quac100;

std::string formatDuration(double ms)
{
    std::ostringstream oss;
    if (ms < 1.0)
    {
        oss << std::fixed << std::setprecision(2) << (ms * 1000.0) << " us";
    }
    else if (ms < 1000.0)
    {
        oss << std::fixed << std::setprecision(2) << ms << " ms";
    }
    else
    {
        oss << std::fixed << std::setprecision(2) << (ms / 1000.0) << " s";
    }
    return oss.str();
}

int main()
{
    std::cout << "QUAC 100 C++ SDK - Hash Operations Example\n";
    std::cout << "==========================================\n\n";
    std::cout.flush();

    try
    {
        std::cout << "Initializing library...\n";
        std::cout.flush();
        Library lib;

        std::cout << "Opening device...\n";
        std::cout.flush();
        auto device = lib.openFirstDevice();
        std::cout << "Device opened successfully.\n\n";
        std::cout.flush();

        std::string testData = "Hello, QUAC 100!";
        Bytes data(testData.begin(), testData.end());
        std::cout << "Test Data: \"" << testData << "\"\n\n";
        std::cout.flush();

        // SHA-2 family
        std::cout << "SHA-2 Family:\n";
        std::cout << "-------------\n";
        std::cout.flush();

        auto sha256 = device.hash().sha256(data);
        std::cout << "SHA-256:  " << utils::toHex(sha256) << "\n";
        std::cout.flush();

        auto sha384 = device.hash().sha384(data);
        std::cout << "SHA-384:  " << utils::toHex(sha384) << "\n";
        std::cout.flush();

        auto sha512 = device.hash().sha512(data);
        std::cout << "SHA-512:  " << utils::toHex(sha512) << "\n\n";
        std::cout.flush();

        // SHA-3 family
        std::cout << "SHA-3 Family:\n";
        std::cout << "-------------\n";
        std::cout.flush();

        auto sha3_256 = device.hash().sha3_256(data);
        std::cout << "SHA3-256: " << utils::toHex(sha3_256) << "\n";
        std::cout.flush();

        auto sha3_512 = device.hash().sha3_512(data);
        std::cout << "SHA3-512: " << utils::toHex(sha3_512) << "\n\n";
        std::cout.flush();

        // SHAKE (extendable output)
        std::cout << "SHAKE (Extendable Output):\n";
        std::cout << "--------------------------\n";
        std::cout.flush();

        auto shake128_32 = device.hash().shake128(data, 32);
        std::cout << "SHAKE128(32):  " << utils::toHex(shake128_32) << "\n";
        std::cout.flush();

        auto shake128_64 = device.hash().shake128(data, 64);
        std::cout << "SHAKE128(64):  " << utils::toHex(shake128_64) << "\n";
        std::cout.flush();

        auto shake256_32 = device.hash().shake256(data, 32);
        std::cout << "SHAKE256(32):  " << utils::toHex(shake256_32) << "\n";
        std::cout.flush();

        auto shake256_64 = device.hash().shake256(data, 64);
        std::cout << "SHAKE256(64):  " << utils::toHex(shake256_64) << "\n\n";
        std::cout.flush();

        // Incremental hashing
        std::cout << "Incremental Hashing:\n";
        std::cout << "--------------------\n";
        std::cout.flush();

        Bytes part1 = {'H', 'e', 'l', 'l', 'o', ',', ' '};
        Bytes part2 = {'Q', 'U', 'A', 'C', ' '};
        Bytes part3 = {'1', '0', '0', '!'};

        auto ctx = device.hash().createContext(HashAlgorithm::SHA256);
        ctx.update(part1);
        ctx.update(part2);
        ctx.update(part3);
        auto incrementalHash = ctx.finalize();

        std::cout << "Incremental: " << utils::toHex(incrementalHash) << "\n";
        std::cout << "One-shot:    " << utils::toHex(sha256) << "\n";
        std::cout << "Match: " << (incrementalHash == sha256 ? "YES" : "NO") << "\n\n";
        std::cout.flush();

        // HMAC
        std::cout << "HMAC:\n";
        std::cout << "-----\n";
        std::cout.flush();

        Bytes key = {'s', 'e', 'c', 'r', 'e', 't', '-', 'k', 'e', 'y'};

        auto hmacSha256 = device.hash().hmac(HashAlgorithm::SHA256, key, data);
        std::cout << "HMAC-SHA256: " << utils::toHex(hmacSha256) << "\n";
        std::cout.flush();

        auto hmacSha512 = device.hash().hmac(HashAlgorithm::SHA512, key, data);
        std::cout << "HMAC-SHA512: " << utils::toHex(hmacSha512) << "\n\n";
        std::cout.flush();

        // HKDF (Key Derivation)
        std::cout << "HKDF Key Derivation:\n";
        std::cout << "--------------------\n";
        std::cout.flush();

        Bytes ikm = device.random().bytes(32); // Input keying material
        Bytes salt = device.random().bytes(16);
        Bytes info = {'a', 'p', 'p', '-', 'k', 'e', 'y'};

        auto derivedKey = device.hash().hkdf(HashAlgorithm::SHA256, ikm, salt, info, 32);
        std::cout << "Derived Key (32 bytes): " << utils::toHex(derivedKey) << "\n";
        std::cout.flush();

        auto derivedKey64 = device.hash().hkdf(HashAlgorithm::SHA256, ikm, salt, info, 64);
        std::cout << "Derived Key (64 bytes): " << utils::toHex(derivedKey64) << "\n\n";
        std::cout.flush();

        // Hash verification example
        std::cout << "Hash Verification Example:\n";
        std::cout << "--------------------------\n";
        std::cout.flush();

        std::string password = "MySecurePassword123";
        Bytes passwordSalt = device.random().bytes(16);

        // Create password hash
        Bytes passwordData(password.begin(), password.end());
        passwordData.insert(passwordData.end(), passwordSalt.begin(), passwordSalt.end());
        auto passwordHash = device.hash().sha256(passwordData);

        std::cout << "Salt: " << utils::toHex(passwordSalt) << "\n";
        std::cout << "Hash: " << utils::toHex(passwordHash) << "\n";
        std::cout.flush();

        // Verify password
        std::string testPassword = "MySecurePassword123";
        Bytes testData2(testPassword.begin(), testPassword.end());
        testData2.insert(testData2.end(), passwordSalt.begin(), passwordSalt.end());
        auto testHash = device.hash().sha256(testData2);

        std::cout << "Verification: " << (testHash == passwordHash ? "MATCH" : "MISMATCH") << "\n\n";
        std::cout.flush();

        // Performance benchmark
        std::cout << "Performance Benchmark (10000 x 1KB):\n";
        std::cout << "------------------------------------\n";
        std::cout.flush();

        Bytes benchData(1024, 0x42);
        const int iterations = 10000;

        auto startTime = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < iterations; ++i)
        {
            device.hash().sha256(benchData);
        }
        auto endTime = std::chrono::high_resolution_clock::now();
        double totalMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        double sha256Rate = (iterations * 1024.0) / (totalMs / 1000.0) / 1024.0 / 1024.0;
        std::cout << "SHA-256:  " << std::fixed << std::setprecision(2)
                  << sha256Rate << " MB/sec\n";
        std::cout.flush();

        startTime = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < iterations; ++i)
        {
            device.hash().sha3_256(benchData);
        }
        endTime = std::chrono::high_resolution_clock::now();
        totalMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        double sha3Rate = (iterations * 1024.0) / (totalMs / 1000.0) / 1024.0 / 1024.0;
        std::cout << "SHA3-256: " << std::fixed << std::setprecision(2)
                  << sha3Rate << " MB/sec\n\n";
        std::cout.flush();

        std::cout << "Hash example completed successfully!\n";
        std::cout.flush();
    }
    catch (const Exception &e)
    {
        std::cerr << "QUAC Error: " << e.what() << " (code: " << e.codeInt() << ")\n";
        std::cerr.flush();
        return 1;
    }
    catch (const std::exception &e)
    {
        std::cerr << "Standard exception: " << e.what() << "\n";
        std::cerr.flush();
        return 1;
    }

    return 0;
}

File: bindings/java/examples/HashExample.java
/*
 * Copyright © 2025 Dyber, Inc. All Rights Reserved.
 * QUAC 100 Post-Quantum Cryptographic Accelerator - Java SDK
 * 
 * Hash Example - Hashing, HMAC, and key derivation operations
 */
package com.dyber.quac100.examples;

import com.dyber.quac100.*;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

/**
 * Example demonstrating hardware-accelerated hashing operations.
 */
public class HashExample {

    public static void main(String[] args) {
        System.out.println("=== QUAC 100 Java SDK Hash Example ===\n");

        try (Library lib = Library.getInstance()) {
            Device device = lib.openFirstDevice();
            Hash hash = device.hash();

            String testData = "Hello, World!";
            byte[] testBytes = testData.getBytes(StandardCharsets.UTF_8);

            // SHA-2 Family
            System.out.println("=== SHA-2 Family ===\n");

            byte[] sha256 = hash.sha256(testBytes);
            System.out.println("SHA-256:  " + Utils.toHex(sha256));

            byte[] sha384 = hash.sha384(testBytes);
            System.out.println("SHA-384:  " + Utils.toHex(sha384));

            byte[] sha512 = hash.sha512(testBytes);
            System.out.println("SHA-512:  " + Utils.toHex(sha512));
            System.out.println();

            // SHA-3 Family
            System.out.println("=== SHA-3 Family ===\n");

            byte[] sha3_256 = hash.sha3_256(testBytes);
            System.out.println("SHA3-256: " + Utils.toHex(sha3_256));

            byte[] sha3_512 = hash.sha3_512(testBytes);
            System.out.println("SHA3-512: " + Utils.toHex(sha3_512));
            System.out.println();

            // SHAKE Extendable Output Functions
            System.out.println("=== SHAKE (Extendable Output) ===\n");

            byte[] shake128_32 = hash.shake128(testBytes, 32);
            System.out.println("SHAKE128 (32 bytes): " + Utils.toHex(shake128_32));

            byte[] shake128_64 = hash.shake128(testBytes, 64);
            System.out.println("SHAKE128 (64 bytes): " + Utils.toHex(shake128_64));

            byte[] shake256_32 = hash.shake256(testBytes, 32);
            System.out.println("SHAKE256 (32 bytes): " + Utils.toHex(shake256_32));

            byte[] shake256_128 = hash.shake256(testBytes, 128);
            System.out.println("SHAKE256 (128 bytes): " +
                    Utils.toHex(Arrays.copyOf(shake256_128, 32)) + "...");
            System.out.println();

            // Incremental hashing
            System.out.println("=== Incremental Hashing ===\n");

            System.out.println("One-shot: hash(\"Hello, World!\")");
            byte[] oneshot = hash.sha256("Hello, World!");
            System.out.println("  Result: " + Utils.toHex(oneshot));

            System.out.println("\nIncremental: update(\"Hello, \") + update(\"World!\")");
            try (Hash.HashContext ctx = hash.createContext(HashAlgorithm.SHA256)) {
                ctx.update("Hello, ");
                ctx.update("World!");
                byte[] incremental = ctx.digest(); // Changed from finalize() to digest()
                System.out.println("  Result: " + Utils.toHex(incremental));

                System.out.println("\nMatches: " +
                        (Arrays.equals(oneshot, incremental) ? "YES ✓" : "NO ✗"));
            }
            System.out.println();

            // HMAC
            System.out.println("=== HMAC (Message Authentication) ===\n");

            byte[] hmacKey = "secret-key-for-hmac".getBytes(StandardCharsets.UTF_8);
            byte[] hmacData = "Data to authenticate".getBytes(StandardCharsets.UTF_8);

            byte[] hmac256 = hash.hmacSha256(hmacKey, hmacData);
            System.out.println("HMAC-SHA256: " + Utils.toHex(hmac256));

            byte[] hmac512 = hash.hmacSha512(hmacKey, hmacData);
            System.out.println("HMAC-SHA512: " + Utils.toHex(hmac512));
            System.out.println();

            // HKDF Key Derivation
            System.out.println("=== HKDF Key Derivation ===\n");

            byte[] ikm = "input-keying-material".getBytes(StandardCharsets.UTF_8);
            byte[] salt = "optional-salt".getBytes(StandardCharsets.UTF_8);
            byte[] info = "context-info".getBytes(StandardCharsets.UTF_8);

            byte[] derived32 = hash.hkdf(HashAlgorithm.SHA256, ikm, salt, info, 32);
            System.out.println("HKDF-SHA256 (32 bytes): " + Utils.toHex(derived32));

            byte[] derived64 = hash.hkdf(HashAlgorithm.SHA256, ikm, salt, info, 64);
            System.out.println("HKDF-SHA256 (64 bytes): " + Utils.toHex(derived64));

            // Derive multiple keys from same material
            System.out.println("\nDeriving encryption and MAC keys:");
            byte[] encKey = hash.hkdf(HashAlgorithm.SHA256, ikm, salt,
                    "encryption".getBytes(), 32);
            byte[] macKey = hash.hkdf(HashAlgorithm.SHA256, ikm, salt,
                    "authentication".getBytes(), 32);
            System.out.println("  Encryption key: " + Utils.toHex(encKey));
            System.out.println("  MAC key:        " + Utils.toHex(macKey));
            System.out.println();

            // Password verification example
            System.out.println("=== Password Verification Example ===\n");

            String password = "MySecurePassword123!";
            byte[] passwordSalt = device.random().bytes(16);

            // Store: hash the password
            byte[] storedHash = hash.hkdf(HashAlgorithm.SHA256,
                    password.getBytes(StandardCharsets.UTF_8),
                    passwordSalt,
                    "password-hash".getBytes(),
                    32);

            System.out.println("Stored salt: " + Utils.toHex(passwordSalt));
            System.out.println("Stored hash: " + Utils.toHex(storedHash));

            // Verify: check entered password
            String enteredPassword = "MySecurePassword123!";
            byte[] verifyHash = hash.hkdf(HashAlgorithm.SHA256,
                    enteredPassword.getBytes(StandardCharsets.UTF_8),
                    passwordSalt,
                    "password-hash".getBytes(),
                    32);

            boolean passwordMatch = Utils.secureCompare(storedHash, verifyHash);
            System.out.println("\nPassword verification: " +
                    (passwordMatch ? "SUCCESS ✓" : "FAILED ✗"));
            System.out.println();

            // Performance benchmark
            System.out.println("=== Performance Benchmark ===\n");

            byte[] benchData = new byte[1024 * 1024]; // 1 MB
            device.random().nextBytes(benchData);

            int iterations = 100;
            long startTime, elapsed;

            // SHA-256
            startTime = System.nanoTime();
            for (int i = 0; i < iterations; i++) {
                hash.sha256(benchData);
            }
            elapsed = System.nanoTime() - startTime;
            double sha256Rate = (iterations * benchData.length) / (elapsed / 1_000_000_000.0) / (1024 * 1024);
            System.out.printf("SHA-256:  %.2f MB/sec%n", sha256Rate);

            // SHA3-256
            startTime = System.nanoTime();
            for (int i = 0; i < iterations; i++) {
                hash.sha3_256(benchData);
            }
            elapsed = System.nanoTime() - startTime;
            double sha3Rate = (iterations * benchData.length) / (elapsed / 1_000_000_000.0) / (1024 * 1024);
            System.out.printf("SHA3-256: %.2f MB/sec%n", sha3Rate);

            device.close();

        } catch (QuacException e) {
            System.err.println("QUAC Error [" + e.getErrorCode() + "]: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("\n=== Example Complete ===");
    }
}